core/chains/evm/gas/rollups/arbitrum_l1_oracle.go

```diff
@@ -0,0 +1,300 @@
+package rollups
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/big"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils"
+
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink/v2/common/client"
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+)
+
+type ArbL1GasOracle interface {
+	L1Oracle
+	GetPricesInArbGas() (perL2Tx uint32, perL1CalldataUnit uint32, err error)
+}
+
+// Reads L2-specific precompiles and caches the l1GasPrice set by the L2.
+type arbitrumL1Oracle struct {
+	services.StateMachine
+	client     l1OracleClient
+	pollPeriod time.Duration
+	logger     logger.SugaredLogger
+	chainType  config.ChainType
+
+	l1GasPriceAddress   string
+	gasPriceMethod      string
+	l1GasPriceMethodAbi abi.ABI
+	l1GasPriceMu        sync.RWMutex
+	l1GasPrice          priceEntry
+
+	l1GasCostAddress   string
+	gasCostMethod      string
+	l1GasCostMethodAbi abi.ABI
+
+	chInitialised chan struct{}
+	chStop        services.StopChan
+	chDone        chan struct{}
+}
+
+const (
+	// ArbGasInfoAddress is the address of the "Precompiled contract that exists in every Arbitrum chain."
+	// https://github.com/OffchainLabs/nitro/blob/f7645453cfc77bf3e3644ea1ac031eff629df325/contracts/src/precompiles/ArbGasInfo.sol
+	ArbGasInfoAddress = "0x000000000000000000000000000000000000006C"
+	// ArbGasInfo_getL1BaseFeeEstimate is the a hex encoded call to:
+	// `function getL1BaseFeeEstimate() external view returns (uint256);`
+	ArbGasInfo_getL1BaseFeeEstimate = "getL1BaseFeeEstimate"
+	// NodeInterfaceAddress is the address of the precompiled contract that is only available through RPC
+	// https://github.com/OffchainLabs/nitro/blob/e815395d2e91fb17f4634cad72198f6de79c6e61/nodeInterface/NodeInterface.go#L37
+	ArbNodeInterfaceAddress = "0x00000000000000000000000000000000000000C8"
+	// ArbGasInfo_getPricesInArbGas is the a hex encoded call to:
+	// `function gasEstimateL1Component(address to, bool contractCreation, bytes calldata data) external payable returns (uint64 gasEstimateForL1, uint256 baseFee, uint256 l1BaseFeeEstimate);`
+	ArbNodeInterface_gasEstimateL1Component = "gasEstimateL1Component"
+	// ArbGasInfo_getPricesInArbGas is the a hex encoded call to:
+	// `function getPricesInArbGas() external view returns (uint256, uint256, uint256);`
+	ArbGasInfo_getPricesInArbGas = "02199f34"
+)
+
+func NewArbitrumL1GasOracle(lggr logger.Logger, ethClient l1OracleClient) ArbL1GasOracle {
```

Comment:
Following the pattern from the OP L1 oracle, should we return `*arbitrumL1Oracle` here instead?
```suggestion
func NewArbitrumL1GasOracle(lggr logger.Logger, ethClient l1OracleClient) *arbitrumL1Oracle {
```

Reply:
Yes, I believe this is best practice